triton/stream.py

```diff
@@ -253,13 +260,227 @@ def checkpoint(self):
+        self.streams = streams
+
+
+    def put(self, **record):
+        return [stream.put(**record) for stream in self.streams]
+
+
+    def put_many(self, records):
+        return [stream.put_many(records) for stream in self.streams]
+
+
+    def build_iterator_from_latest(self):
+        return CompositeStream.Iterator(self.streams)
+
+
+    def build_iterator_from_checkpoint(self, checkpoints):
+        return CompositeStream.Iterator(self.streams, checkpoints)
+
+
+class GCPStream(Stream):
+
+
+    # Pubsub limits as defined https://cloud.google.com/pubsub/quotas.
+    BATCH_MAX_MSGS = 1000
+
+    class Iterator(object):
+
+
+        def __init__(self, stream, subscription_id=None):
+            self.stream = stream
+            self.subscription_id = subscription_id or 'trition-{}'.format(uuid.uuid4().hex)
+            self.subscription = stream.topic.subscription(self.subscription_id)
+            if subscription_id is None:
+                # If the subscription_id is provided, then
+                # it is assumed we are resuming from a previously
+                # active, named subscription.
+                self.subscription.create()
+
+            self.prefetch = []
+            self.prefetch_ack_ids = []
+            self.pending_acks = []
+            self.first_next = True
+
+
+        def __iter__(self):
+            return self
+
+
+        def next(self):
+            return self.__next__()
+
+
+        def __next__(self):
+            if self.first_next is False:
+                self.pending_acks.append(self.prefetch_ack_ids.pop(0))
+            else:
+                self.first_next = False
+
+            if len(self.prefetch) == 0:
+                if len(self.pending_acks) > 0:
+                    self.subscription.acknowledge(self.pending_acks)
+                    self.pending_acks = []
+
+            pubsub_messages = self.subscription.pull(return_immediately=False)
+            for ack_id, raw_message in pubsub_messages:
+                self.prefetch_ack_ids.append(ack_id)
+                self.prefetch.append(self.stream.decode(raw_message))
+
+            if len(self.prefetch) == 0:
+                raise StopIteration
+
+            return self.prefetch.pop(0)
+
+
+    def __init__(self, project, topic, private_key_file):
+        self.http = None
+        self.credentials = None
+
+        # In order to support emulation we either pass
+        # user provided credentials into pubsub.Client or a
+        # simple http client.
+        if private_key_file:
+            self.credentials = service_account.Credentials.from_service_account_file(
+                private_key_file)
+        else:
+            self.http = requests.Session()
+
+        self.client = pubsub.Client(project, credentials=self.credentials, _http=self.http)
+        self.project = project
+        self.topic_name = topic
+        self.topic = self.client.topic(self.topic_name)
+
+        super(GCPStream, self).__init__()
```

Comment:
This is just an opinion, so feel free to disregard:

As a pattern, I think it's better to do the super initialization first. The superclass(es) shouldn't need to have any attributes initialized by a subclass prior to itself.